openvdb_points/tools/AttributeArrayString.cc

```diff
@@ -0,0 +1,322 @@
+// and license notice and the following restrictions and disclaimer.
+//
+// *     Neither the name of Double Negative Visual Effects nor the names
+// of its contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// IN NO EVENT SHALL THE COPYRIGHT HOLDERS' AND CONTRIBUTORS' AGGREGATE
+// LIABILITY FOR ALL CLAIMS REGARDLESS OF THEIR BASIS EXCEED US$250.00.
+//
+///////////////////////////////////////////////////////////////////////////
+//
+/// @file AttributeArrayString.cc
+///
+/// @authors Dan Bailey
+
+
+#include <openvdb_points/tools/AttributeArrayString.h>
+
+#include <openvdb/Metadata.h>
+#include <openvdb/metadata/MetaMap.h>
+
+#include <boost/algorithm/string/predicate.hpp> // startswith
+
+#include <sstream>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+
+namespace {
+
+    bool isStringMeta(const Name& key, const Metadata::Ptr& meta)
+    {
+        // ensure the metadata is StringMetadata
+        if (meta->typeName() != "string")           return false;
+        // string attribute metadata must have a key that starts "string:"
+        if (!boost::starts_with(key, "string:"))    return false;
+
+        return true;
+    }
+
+    Name getStringKey(const StringType index)
+    {
+        std::stringstream ss;
+        ss << "string:" << (index - 1);
+        return ss.str();
+    }
+
+    StringType getStringIndex(const Name& key)
+    {
+        Name indexStr = key.substr(7, key.size() - 7);
+
+        // extract the index as an unsigned integer
+        std::istringstream indexSS(indexStr);
+        Index index;
+        indexSS >> index;
+
+        return (index + 1);
+    }
+
+} // namespace
+
+
+////////////////////////////////////////
+
+
+// StringMetaInserter implementation
+
+
+StringMetaInserter::StringMetaInserter(MetaMap& metadata)
+    : mMetadata(metadata)
+{
+    // populate the cache
+    resetCache();
+}
+
+
+void StringMetaInserter::insert(const Name& name)
+{
+    // name already exists, so do nothing
+
+    if (std::find(mValues.begin(), mValues.end(), name) != mValues.end())  return;
```

Comment:
I wouldn't expect this to scale well to cope with large numbers of unique strings (for instance, asset uuids for point-instanced geometry workflows). Would switching to a hashing data structure (e.g. std::set) be viable for mValues? It doesn't look like you are relying on the order of mValues anywhere, so this should hopefully allow you to remove a std::sort and replace the linear search with a cheaper set lookup...